#!/bin/bash

if [ -f "${RESULTS_DIR}/${NAME}_total_extracted_1.fastq" ] || [ -f "${RESULTS_DIR}/${NAME}_total_extracted_2.fastq" ];then
	echo "READS are already extracted."
	exit 0
fi

module load samtools/1.10

# Extract reads that were aligned to HLA loci (HLA-A, B, C, DM, DO, DP, DQ, DR, E, F, G, H, J, K, L, P, V, MIC, and TAP)
REGIONS='6:29907037-29915661 6:31319649-31326989 6:31234526-31241863 6:32914391-32922899 6:32900406-32910847 6:32969960-32979389 6:32778540-32786825 6:33030346-33050555 6:33041703-33059473 6:32603183-32613429 6:32707163-32716664 6:32625241-32636466 6:32721875-32733330 6:32405619-32414826 6:32544547-32559613 6:32518778-32554154 6:32483154-32559613 6:30455183-30463982 6:29689117-29699106 6:29792756-29800899 6:29793613-29978954 6:29855105-29979733 6:29892236-29899009 6:30225339-30236728 6:31369356-31385092 6:31460658-31480901 6:29766192-29772202 6:32810986-32823755 6:32779544-32808599 6:29756731-29767588'
REGIONS_100='6:29906937-29915761 6:31319549-31327089 6:31234426-31241963 6:32914291-32922999 6:32900306-32910947 6:32969860-32979489 6:32778440-32786925 6:33030246-33050655 6:33041603-33059573 6:32603083-32613529 6:32707063-32716764 6:32625141-32636566 6:32721775-32733430 6:32405519-32414926 6:32544447-32559713 6:32518678-32554254 6:32483054-32559713 6:30455083-30464082 6:29689017-29699206 6:29792656-29800999 6:29793513-29979054 6:29855005-29979833 6:29892136-29899109 6:30225239-30236828 6:31369256-31385192 6:31460558-31481001 6:29766092-29772302 6:32810886-32823855 6:32779444-32808699 6:29756631-29767688'
samtools view -bu "$PATH_TO_BAM" $REGIONS_100 |
samtools collate -f -O -u -r 1000000 /dev/stdin |
samtools fastq -0 /dev/null -1 "${RESULTS_DIR}/${NAME}_hla_extracted_1.fastq" -2 "${RESULTS_DIR}/${NAME}_hla_extracted_2.fastq" -s /dev/null

# Extract unmapped reads
samtools view -bu -f 12 "$PATH_TO_BAM" |
samtools collate -f -O -u -r 1000000 /dev/stdin |
samtools fastq -0 /dev/null -1 "${RESULTS_DIR}/${NAME}_unmapped_extracted_1.fastq" -2 "${RESULTS_DIR}/${NAME}_unmapped_extracted_2.fastq" -s /dev/null

# Merge HLA-originating and mapped reads
cat "${RESULTS_DIR}/${NAME}_hla_extracted_1.fastq" "${RESULTS_DIR}/${NAME}_unmapped_extracted_1.fastq" > "${RESULTS_DIR}/${NAME}_total_extracted_1.fastq"
cat "${RESULTS_DIR}/${NAME}_hla_extracted_2.fastq" "${RESULTS_DIR}/${NAME}_unmapped_extracted_2.fastq" > "${RESULTS_DIR}/${NAME}_total_extracted_2.fastq"

